engine: add byte limit to MVCCScan

tbg opened the pull request:
A fledgling step towards #19721 is allowing incoming KV requests to
bound the size of the response in terms of bytes rather than rows.
This commit adds a TargetBytes field to MVCCScanOptions to address
this need: scans stop once the size of the result meets or exceeds the
threshold (at least one key will be added, regardless of its size),
and returns a ResumeSpan as appropriate.

The classic example of the problem this addresses is a table in which
each row is, say, ~1mb in size. A full table scan will currently fetch
data from KV in batches of [10k], causing at least 10GB of data held in
memory at any given moment. This sort of thing does happen in practice;
we have a long-failing roachtest #33660 because of just that, and
anecdotally OOMs in production clusters are with regularity caused by
individual queries consuming excessive amounts of memory at the KV
level.

Plumbing this limit into a header field on BatchRequest and down to the
engine level will allow the batching in [10k] to become byte-sized in
nature, thus avoiding one obvious source OOMs. This doesn't solve #19721
in general (many such requests could work together to consume a lot of
memory after all), but it's a sane baby step that might just avoid a
large portion of OOMs already.

[10k]: https://github.com/cockroachdb/cockroach/blob/0a658c19cd164e7c021eaff7f73db173f0650e8c/pkg/sql/row/kv_batch_fetcher.go#L25-L29

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44339)
<!-- Reviewable:end -->

petermattis reviewed (commented):
Both the KVs and Bytes response formats go through the same `mvccScanner/pebbleMVCCScanner` code paths. We always scan to the Bytes format, and then convert to KVs if necessary. I think you only have to do this work twice, not four times. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-:-LzMkf8h1mFKnV2FEeDy:b-j5znx6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @tbg)

___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzMl3xPC2ZNrUDcoO1Y:-LzMl3xQ1u4JMXNWXX9E:b50zu7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a8d69052d3e90b80f572b49e7c33f15c2a1b144/pkg/storage/engine/pebble_mvcc_scanner.go#L536)):*
> ```Go
> 			// pretend we reached maxKeys. (The resume key computation requires
> 			// fetching an additional key after that, so trying to exit the
> 			// iterations based on size alone is not trivial).
> ```

That is a bit hacky, though it seems safer than trying to extend the resume key logic.


<!-- Sent from Reviewable.io -->

itsbilal commented:
Looks good to me so far. It may not be too much more effort to sprinkle in `if p.result.bytes >= p.maxBytes && p.maxBytes > 0` checks in all 3-4 of the cases where we already have `maxKeys` checks, but given how this is a hot path it might only make things less performant than just updating `maxKeys`, so I'm fine with this approach. Possibly with a comment on the dual-use of `maxKeys` where it is defined.

And yes, like Peter said, you'll have to make the same change only once more, in `mvcc.h` for the rocksdb mvcc scanner. The KV/bytes cases wrap around these two implementations.

You should also be able to use Raphael's new datadriven-based test framework to quickly write up some tests for this. Take a look at `TestMVCCHistories` and the giant comment there for how to use it.

tbg commented:
Before I write tests, any concerns/ideas around naming this? We're not
really limiting the max size strictly speaking because we "overshoot". OTOH
this behavior makes sense (you wouldn't want to return zero results when
there's a resume key). There must be a "right" name for this sort of thing.

On Fri, Jan 24, 2020 at 4:26 PM Bilal Akhtar <notifications@github.com>
wrote:

> Looks good to me so far. It may not be too much more effort to sprinkle in if
> p.result.bytes >= p.maxBytes && p.maxBytes > 0 checks in all 3-4 of the
> cases where we already have maxKeys checks, but given how this is a hot
> path it might only make things less performant than just updating maxKeys,
> so I'm fine with this approach. Possibly with a comment on the dual-use of
> maxKeys where it is defined.
>
> And yes, like Peter said, you'll have to make the same change only once
> more, in mvcc.h for the rocksdb mvcc scanner. The KV/bytes cases wrap
> around these two implementations.
>
> You should also be able to use Raphael's new datadriven-based test
> framework to quickly write up some tests for this. Take a look at
> TestMVCCHistories and the giant comment there for how to use it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/44339?email_source=notifications&email_token=ABGXPZBCCOVB5RH2UGLKBBDQ7MCDBA5CNFSM4KLFRT4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ3EMMA#issuecomment-578176560>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGXPZEVTZH7ZMVPU2WTAXDQ7MCDBANCNFSM4KLFRT4A>
> .
>

petermattis commented:
`TargetBytes`? You can both overshoot and undershoot a target.

On Mon, Jan 27, 2020 at 7:38 AM Tobias Grieger <notifications@github.com>
wrote:

> Before I write tests, any concerns/ideas around naming this? We're not
> really limiting the max size strictly speaking because we "overshoot". OTOH
> this behavior makes sense (you wouldn't want to return zero results when
> there's a resume key). There must be a "right" name for this sort of thing.
>
> On Fri, Jan 24, 2020 at 4:26 PM Bilal Akhtar <notifications@github.com>
> wrote:
>
> > Looks good to me so far. It may not be too much more effort to sprinkle
> in if
> > p.result.bytes >= p.maxBytes && p.maxBytes > 0 checks in all 3-4 of the
> > cases where we already have maxKeys checks, but given how this is a hot
> > path it might only make things less performant than just updating
> maxKeys,
> > so I'm fine with this approach. Possibly with a comment on the dual-use
> of
> > maxKeys where it is defined.
> >
> > And yes, like Peter said, you'll have to make the same change only once
> > more, in mvcc.h for the rocksdb mvcc scanner. The KV/bytes cases wrap
> > around these two implementations.
> >
> > You should also be able to use Raphael's new datadriven-based test
> > framework to quickly write up some tests for this. Take a look at
> > TestMVCCHistories and the giant comment there for how to use it.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/44339?email_source=notifications&email_token=ABGXPZBCCOVB5RH2UGLKBBDQ7MCDBA5CNFSM4KLFRT4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ3EMMA#issuecomment-578176560
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABGXPZEVTZH7ZMVPU2WTAXDQ7MCDBANCNFSM4KLFRT4A
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/44339?email_source=notifications&email_token=ABPJ753QRCKPLT46LYKJCE3Q73IWHA5CNFSM4KLFRT4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ7LDQI#issuecomment-578728385>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPJ7566AG22B5I67DJPAZ3Q73IWHANCNFSM4KLFRT4A>
> .
>

tbg commented:
> It may not be too much more effort to sprinkle in if p.result.bytes >= p.maxBytes && p.maxBytes > 0

@itsbilal that's what I did initially, but it was... hard, because `maxKeys` has this statemachine-like property where the behavior is different exactly when `result.count == maxKeys` (but then you need to be able to get the resume key). A cleaner way to structure this would be nice, though I'm just going to go with the current approach to avoid getting sidetracked.

tbg commented:
This PR is now ready for review in earnest.

tbg commented on pkg/storage/engine/testdata/mvcc_histories/target_bytes:
```diff
@@ -0,0 +1,58 @@
+## Test opts.TargetBytes.
```
The testing seems a little on the light side to me, accepting suggestions on what else to test.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though @itsbilal should give this a review as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-:-Lzqj11iARIyuT824vlw:b-98gv1z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @tbg)

___
*[c-deps/libroach/chunked_buffer.cc, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzqiGZJ4Ow3lLXvLf_D:-LzqiGZJ4Ow3lLXvLf_E:b4t8w3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/c-deps/libroach/chunked_buffer.cc#L32)):*
> ```cpp
>   put(value.data(), value.size(), 0);
>   count_++;
>   bytes_ += 8 + key.size() + value.size(); // see (*pebbleResults).put
> ```

`s/8/sizeof(size_buf)/g`
___
*[pkg/storage/engine/testdata/mvcc_histories/target_bytes, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzqG2QwjqI430RnmfKT-r2-1:-Lzqhr2YC3I7DmaZU74X:b-cuwbcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/pkg/storage/engine/testdata/mvcc_histories/target_bytes#L1)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The testing seems a little on the light side to me, accepting suggestions on what else to test.
</blockquote></details>

Reverse scans?

Perhaps a test where keys are skipped. That is, show that the target bytes is only affected by keys that are added to the result, not keys that are skipped because they are too new.
___
*[pkg/storage/engine/testdata/mvcc_histories/target_bytes, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-Lzqib-1DU_48K6Y988C:-Lzqib-1DU_48K6Y988D:bfgd653) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/pkg/storage/engine/testdata/mvcc_histories/target_bytes#L7)):*
> ```
>   put      k=a v=abcdef
>   put      k=c v=ghijkl
>   put      k=e v=mnopqr
> ```

Might want to vary the value size. Right now, every key and value has the same size. This does make the results line up nicely, but is perhaps a bit too uniform for testing purposes.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks!

> Reviewed 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-:-LzrU2Ur4boOvtr93gsz:brfqlux)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @tbg)

___
*[c-deps/libroach/mvcc.h, line 728 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzrLJ_RAqtvu3OD7vim:-LzrLJ_RAqtvu3OD7vin:b-8dp5dh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/c-deps/libroach/mvcc.h#L728)):*
> ```objc
>   const rocksdb::Slice end_key_;
>   int64_t max_keys_;
>   const int64_t target_bytes_; // see pebbleMVCCScanner
> ```

Might be worth saying `see MVCCScanOptions` instead, since the big comment explaining this is there?
___
*[pkg/storage/engine/mvcc_history_test.go, line 700 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzrTZTp7prypWVximnx:-LzrTZTp7prypWVximny:bn8759) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/pkg/storage/engine/mvcc_history_test.go#L700)):*
> ```Go
> 		max = int64(imax)
> 	}
> 	if key := "tb"; e.hasArg(key) {
> ```

This works but I'm wondering if we can use a more descriptive arg name for this, given my first instinct was "tb? sounds like TomBstones". `targetbytes` probably works better and is less ambiguous.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 539 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzrRc6-57ZtiuRn2-rx:-LzrRc60Du4ipQoEWoiR:bh6nlx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/pkg/storage/engine/pebble_mvcc_scanner.go#L539)):*
> ```Go
> 			// number of keys.
> 			//
> 			// TODO(bilal): see if this can be implemented more transparently.
> ```

:+1: on the todo
___
*[pkg/storage/engine/testdata/mvcc_histories/target_bytes, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzqG2QwjqI430RnmfKT-r2-1:-LzrS18z3w2afXlKON8G:bqpzmf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/pkg/storage/engine/testdata/mvcc_histories/target_bytes#L1)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Reverse scans?

Perhaps a test where keys are skipped. That is, show that the target bytes is only affected by keys that are added to the result, not keys that are skipped because they are too new.
</blockquote></details>

:+1: on the reverse scan suggestion.

Also, another possibility is zero-sized values (basically tombstones) and scans with tombstones=true


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTRs! I souped up the tests with reverse iteration and tombstones (and the combination of both). Everything worked as expected.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-:-M-9EKGX8fwcPIR4vJMz:b-gg67hq)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @petermattis)

___
*[c-deps/libroach/chunked_buffer.cc, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-LzqiGZJ4Ow3lLXvLf_D:-M-9EJUj1-Gf_L3HjEeo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/c-deps/libroach/chunked_buffer.cc#L32)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/8/sizeof(size_buf)/g`
</blockquote></details>

Done.
___
*[pkg/storage/engine/testdata/mvcc_histories/target_bytes, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44339#-Lzqib-1DU_48K6Y988C:-M-9CJ5IE7F90Ji8ZAiK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4346dac4a3a0e6afe719ba05c996461918730d82/pkg/storage/engine/testdata/mvcc_histories/target_bytes#L7)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Might want to vary the value size. Right now, every key and value has the same size. This does make the results line up nicely, but is perhaps a bit too uniform for testing purposes.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg committed:
engine: add byte limit to MVCCScan

A fledgling step towards #19721 is allowing incoming KV requests to
bound the size of the response in terms of bytes rather than rows.
This commit adds a TargetBytes field to MVCCScanOptions to address
this need: scans stop once the size of the result meets or exceeds the
threshold (at least one key will be added, regardless of its size),
and returns a ResumeSpan as appropriate.

The classic example of the problem this addresses is a table in which
each row is, say, ~1mb in size. A full table scan will currently fetch
data from KV in batches of [10k], causing at least 10GB of data held in
memory at any given moment. This sort of thing does happen in practice;
we have a long-failing roachtest #33660 because of just that, and
anecdotally OOMs in production clusters are with regularity caused by
individual queries consuming excessive amounts of memory at the KV
level.

Plumbing this limit into a header field on BatchRequest and down to the
engine level will allow the batching in [10k] to become byte-sized in
nature, thus avoiding one obvious source OOMs. This doesn't solve #19721
in general (many such requests could work together to consume a lot of
memory after all), but it's a sane baby step that might just avoid a
large portion of OOMs already.

[10k]: https://github.com/cockroachdb/cockroach/blob/0a658c19cd164e7c021eaff7f73db173f0650e8c/pkg/sql/row/kv_batch_fetcher.go#L25-L29

Release note: None

tbg commented:
bors r=petermattis,itsbilal

jordanlewis commented:
woo!
